src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingServiceFactory.cs

```diff
@@ -1,29 +1,105 @@
     {
-        private readonly IServiceProvider _serviceProvider;
-        private IDocumentTrackingService _singleton;
+        private readonly VisualStudioActiveDocumentTracker _activeDocumentTracker;
 
         [ImportingConstructor]
-        public VisualStudioDocumentTrackingServiceFactory(SVsServiceProvider serviceProvider)
+        public VisualStudioDocumentTrackingServiceFactory(VisualStudioActiveDocumentTracker activeDocumentTracker)
         {
-            _serviceProvider = serviceProvider;
+            _activeDocumentTracker = activeDocumentTracker;
         }
 
         public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
         {
-            return _singleton ?? (_singleton = new VisualStudioDocumentTrackingService(_serviceProvider));
+            return new VisualStudioDocumentTrackingService(_activeDocumentTracker, workspaceServices.Workspace);
+        }
+
+        private class VisualStudioDocumentTrackingService : IDocumentTrackingService
+        {
+            private readonly VisualStudioActiveDocumentTracker _activeDocumentTracker;
+            private readonly Workspace _workspace;
+
+            public VisualStudioDocumentTrackingService(VisualStudioActiveDocumentTracker activeDocumentTracker, Workspace workspace)
+            {
+                _activeDocumentTracker = activeDocumentTracker;
+                _workspace = workspace;
+            }
+
+            private object _gate = new object();
+            private int _subscriptions = 0;
+            private event EventHandler<DocumentId> _activeDocumentChangedEventHandler;
+
+            public event EventHandler<DocumentId> ActiveDocumentChanged
+            {
+                add
+                {
+                    lock (_gate)
+                    {
+                        _subscriptions++;
+
+                        if (_subscriptions == 1)
+                        {
+                            _activeDocumentTracker.DocumentsChanged += ActiveDocumentTracker_DocumentsChanged;
+                        }
+
+                        _activeDocumentChangedEventHandler += value;
+                    }
+                }
+
+                remove
+                {
+                    lock (_gate)
+                    {
+                        _subscriptions--;
+
+                        if (_subscriptions == 0)
+                        {
+                            _activeDocumentTracker.DocumentsChanged -= ActiveDocumentTracker_DocumentsChanged;
+                        }
+
+                        _activeDocumentChangedEventHandler -= value;
+                    }
+                }
+            }
+
+            private void ActiveDocumentTracker_DocumentsChanged(object sender, EventArgs e)
+            {
+                _activeDocumentChangedEventHandler?.Invoke(this, GetActiveDocument());
+            }
+
+            public event EventHandler<EventArgs> NonRoslynBufferTextChanged
```

Comment:
def interested in why we need this.

Reply:
We use it to let the solution crawler back off. We have logic to back of if we are editing our files (to avoid us causing lots of GC for example), but if say you're editing a Readme file or something else we don't want to be making that terrible.

Reply:
@heejaechang and I wouldn't necessarily be against deleting this, but at the same time the optimization does make sense. If there was an easy way to subscribe to all text buffer changes for all languages we'd just do that, use that to drive the solution crawler, and be done with it. But there's not, hence the silliness.

Reply:
No, that makes a lot of sense.  Sadly, this also would seem to be relevant for people doing *anything* in VS beyond just editing. i.e. if i'm working with a designer, it woudl be good for roslyn to back off.  Unfortunately, i don't imagine there's any good service that VS exposes to give a generalized mechanism for this.

Reply:
Yeah, we don't have a good service for that. The editor team is looking at some new APIs for tracking buffer/view lifetimes, so something might be useful when that comes up.